### PR TITLE
Citizen loses xp if his new job != old job

### PR DESCRIPTION
**Pull request by Raycoms**
# Changes proposed in this pull request:
- Citizens will lose their xp if hired in a different job.
-
-

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/colony/CitizenData.java [hunk 1]
```diff
@@ -619,6 +619,10 @@ public int getLevel()
      */
     public void setLevel(final int lvl)
     {
+        if(lvl == 0)
+        {
+            this.experience = 0;
```
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/colony/CitizenData.java]: Can we not just make a second method named `resetExperience` that sets level and experience to 0?

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java [hunk 2]
```diff
@@ -73,14 +73,21 @@ public void setWorker(final CitizenData citizen)
...
 
         worker = citizen;
 
         // If we set a worker, inform it of such
         if (worker != null)
         {
+            EntityCitizen tempCitizen = citizen.getCitizenEntity();
+            if(!tempCitizen.getLastJob().equals(getJobName()))
+            {
+                citizen.setLevel(0);
```
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java]: then we should here use the proposed `resetExperience`

**Review by marvin-bitterlich — CHANGES_REQUESTED** [review]
For Code clarity we should change this.
Looking good otherwise!

File: src/main/java/com/minecolonies/coremod/colony/CitizenData.java (hunk 1 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/CitizenData.java]: done, look again =D

Comment (marvin-bitterlich) [reply src/main/java/com/minecolonies/coremod/colony/CitizenData.java]: can you remove the `if(lvl == 0)` part?

**Review by marvin-bitterlich — APPROVED** [review]

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java [hunk 3]
```diff
@@ -73,14 +73,21 @@ public void setWorker(final CitizenData citizen)
...
             worker.setWorkBuilding(null);
+            tempCitizen.setLastJob(getJobName());
         }
 
         worker = citizen;
 
         // If we set a worker, inform it of such
         if (worker != null)
         {
+            EntityCitizen tempCitizen = citizen.getCitizenEntity();
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java]: MINOR Local variable 'tempCitizen' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java [hunk 4]
```diff
@@ -73,14 +73,21 @@ public void setWorker(final CitizenData citizen)
         // If we have a worker, it no longer works here
         if (worker != null)
         {
+            EntityCitizen tempCitizen = worker.getCitizenEntity();
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java]: MINOR Local variable 'tempCitizen' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 2 issues
* MINOR 2 minor

Watch the comments in this conversation to review them.

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java [hunk 5]
```diff
@@ -73,14 +73,21 @@ public void setWorker(final CitizenData citizen)
...
             worker.setWorkBuilding(null);
+            tempCitizen.setLastJob(getJobName());
         }
 
         worker = citizen;
 
         // If we set a worker, inform it of such
         if (worker != null)
         {
+            final EntityCitizen tempCitizen = citizen.getCitizenEntity();
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java]: MINOR Local variable 'tempCitizen' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java [hunk 6]
```diff
@@ -73,14 +73,21 @@ public void setWorker(final CitizenData citizen)
         // If we have a worker, it no longer works here
         if (worker != null)
         {
+            final EntityCitizen tempCitizen = worker.getCitizenEntity();
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java]: MINOR Local variable 'tempCitizen' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java (hunk 5 above)
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java:86]: MINOR Local variable 'tempCitizen' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java (hunk 6 above)
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java:76]: MINOR Local variable 'tempCitizen' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 2 issues
* MINOR 2 minor

Watch the comments in this conversation to review them.

**Comment by Raycoms** [comment]
can we @Kostronor

**Comment by marvin-bitterlich** [comment]
LGTM!

[Approved with PullApprove](https://pullapprove.com/Minecolonies/minecolonies/pull-request/337/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Minecolonies/minecolonies)

**Pull request merged**